### PR TITLE
[64] Search documents using data labels

### LINKED ISSUE #64: Generating searches by existing labels
This issue should:
- Produce a list of candidate GtR projects given a query of research topics
- Produce a list of Crunchbase organisations given a query of industry labels

Inputs:
- Crunchbase and GtR labels

Outputs:
- A list of GtR project ids, and Crunchbase organisation ids pertaining to the cross-dataset categories of interest

### PR DESCRIPTION
**Pull request by beingkk**
Closes #64 

Implements project and company querying using the existing GtR and Crunchbase data labels (categories). The results are output in the same format as in the case of keyword queries (ie, a dataframe with a column for each query where in this case the query is the specified label/category).

Usage examples are shown in `examples/07_query_terms.py` and in the [README](https://github.com/nestauk/innovation_sweet_spots/blob/64_label_searching/innovation_sweet_spots/analysis/README.md#query_categories).

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

### CODE REVIEW

**Comment by beingkk** [comment]
Hi @Jack-0-0 , this PR might be an interesting one to check out: I tried writing a couple tests using mock decorators and "side effects".

File: innovation_sweet_spots/analysis/query_categories.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        # Keep track of items in any of the provided categories
+        items_in_any_category = items_in_any_category | set(items_in_category)
+    # Add a column to mark items which are in any category
+    matches["any_category"] = matches["id"].isin(items_in_any_category)
+    # Return all items or only items which are in the category
+    if return_only_matches:
+        return matches.query("any_category == True").reset_index(drop=True)
+    else:
+        return matches
+    return matches
```
Comment (beingkk) [inline innovation_sweet_spots/analysis/query_categories.py]:
```suggestion
```

File: innovation_sweet_spots/analysis/query_categories.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        )
+        return []
+    else:
+        # Make sure that there is only one unique GtR research topic id
+        assert (
+            len(topic_id_df) == 1
+        ), "There is more than one topic id with the same name: Something wrong with GtrWrangler.gtr_topics"
+        topic_id = topic_id_df.id.iloc[0]
+        # Find the project ids in the specified category
+        return GtR.link_gtr_topics.query("id == @topic_id").project_id.to_list()
```
Comment (Jack-0-0) [inline innovation_sweet_spots/analysis/query_categories.py:96]: What does the `@` in the query do?

File: innovation_sweet_spots/analysis/wrangling_utils.py [hunk 3]
```diff
@@ -409,6 +409,10 @@ def gtr_topics(self):
             self._gtr_topics = gtr.get_gtr_topics().rename(columns={"text": "topic"})
         return self._gtr_topics
 
+    @property
+    def gtr_topics_list(self):
+        return sorted(self.gtr_topics.topic.to_list())
```
Comment (Jack-0-0) [inline innovation_sweet_spots/analysis/wrangling_utils.py]: There's no docstring for this function (maybe not needed, as the function is straightforward 🙂)

File: innovation_sweet_spots/tests/test_query_categories.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+)
+def test_query_gtr_categories(
+    mock_is_gtr_project_in_category, mock_initialise_gtr_id_table
+):
+
+    mock_is_gtr_project_in_category.return_value = []
+    mock_is_gtr_project_in_category.side_effect = category_check_side_effect
+    mock_initialise_gtr_id_table.return_value = pd.DataFrame(
+        {"id": ["id_1", "id_2", "id_3"]}
+    )
```
Comment (Jack-0-0) [inline innovation_sweet_spots/tests/test_query_categories.py]: Here, you can move the return values and side effects into `@mock.patch`
```suggestion
@mock.patch(
    "innovation_sweet_spots.analysis.query_categories.initialise_gtr_id_table",
    return_value=pd.DataFrame({"id": ["id_1", "id_2", "id_3"]}),
)
@mock.patch(
    "innovation_sweet_spots.analysis.query_categories.is_gtr_project_in_category",
    return_value=[],
    side_effect=category_check_side_effect,
)
def test_query_gtr_categories(
    mock_is_gtr_project_in_category, mock_initialise_gtr_id_table
):
```

File: innovation_sweet_spots/tests/test_query_categories.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+)
+def test_query_cb_categories(
+    mock_is_cb_organisation_in_category, mock_initialise_cb_id_table
+):
+
+    mock_is_cb_organisation_in_category.return_value = []
+    mock_is_cb_organisation_in_category.side_effect = category_check_side_effect
+    mock_initialise_cb_id_table.return_value = pd.DataFrame(
+        {"id": ["id_1", "id_2", "id_3"]}
+    )
```
Comment (Jack-0-0) [inline innovation_sweet_spots/tests/test_query_categories.py]: Same as above, you can move the return values and side effects into `@mock.patch`

**Review by Jack-0-0 — COMMENTED** [review]
The tests are done in a nice and simple way, thanks for showing me this @beingkk.
The notebook runs as expected. I’ve added some comments on specific lines.

File: innovation_sweet_spots/analysis/query_categories.py (hunk 2 above)
Comment (beingkk) [reply innovation_sweet_spots/analysis/query_categories.py:96]: Hey, `@` allows `.query()` to use variables from the environment (in this case `topic_id`)

File: innovation_sweet_spots/analysis/wrangling_utils.py (hunk 3 above)
Comment (beingkk) [reply innovation_sweet_spots/analysis/wrangling_utils.py]: Thanks will add a docstring for a good measure
```suggestion
    def gtr_topics_list(self):
        """Returns a sorted list of GtR research topics (categories)"""
        return sorted(self.gtr_topics.topic.to_list())
```

File: innovation_sweet_spots/tests/test_query_categories.py (hunk 4 above)
Comment (beingkk) [reply innovation_sweet_spots/tests/test_query_categories.py]: Nice, that is neater - thank you!

File: innovation_sweet_spots/tests/test_query_categories.py (hunk 5 above)
Comment (beingkk) [inline innovation_sweet_spots/tests/test_query_categories.py]:
```suggestion
@mock.patch(
    "innovation_sweet_spots.analysis.query_categories.initialise_cb_id_table",
    return_value = pd.DataFrame({"id": ["id_1", "id_2", "id_3"]}
)
@mock.patch(
    "innovation_sweet_spots.analysis.query_categories.is_cb_organisation_in_category",
    return_value = [],
    side_effect = category_check_side_effect
)
def test_query_cb_categories(
    mock_is_cb_organisation_in_category, mock_initialise_cb_id_table
):
```

Comment (beingkk) [reply innovation_sweet_spots/tests/test_query_categories.py]: Thanks!

File: innovation_sweet_spots/analysis/query_categories.py (hunk 2 above)
Comment (Jack-0-0) [reply innovation_sweet_spots/analysis/query_categories.py:96]: Ah, thanks for the explanation, I hadn't seen it before 🙂  I usually use f-strings there

**Pull request merged**